### ISSUE TITLE
character bounds for connected font is not calculate correctly

### ISSUE THREAD

**Issue by minzojian** [kangax/fabric.js#4815]
character bounds for connected font is not calculate correctly

## Version
2.2.0

## Test Case
http://fabricjs.com/loadfonts

## Steps to reproduce
visit the offical demo, and chose a connected font, like the second font (**Pacifico**) in the dropdown list
type some **fffffff** in the textbox, rember zoom in the textbox, you can see the result clear

## Expected Behavior
connected font is cool, but character bounds calculate not correctly, it effect the cusor location and the selection react render

## Actual Behavior
![image](https://user-images.githubusercontent.com/361781/37381891-bb1ca45c-277a-11e8-83cc-35c5f2d2727f.png)

i select all text, but as you can see, it is not include all characters

**Comment by asturur** [comment]
https://github.com/kangax/fabric.js/pull/4772/files

I thought this was the root cause, but i see it needs more inspection.
How does the cursor position itself? Good or bad?

**Comment by blobinabottle** [comment]
Bad position cursor on this specific font. :-/

**Comment by asturur** [comment]
in my app works fine, did not try on kitchensink yet.

**Comment by blobinabottle** [comment]
You need to make the text quite long to see the effect.

**Comment by jilster** [comment]
I am quite sure this problem comes from the font itself and not from Fabric. It seems to only happen under very specific circumstances. Only in the Pacifico font, with the letter 'f', and only in ligature mode.  If a font supports ligatures (connected characters), each character has 2 sets of info. 1 for standalone mode and 1 for ligature mode. Specifically the ligature mode of the letter 'f' seems te be off.

If you go to https://fonts.google.com/specimen/Pacifico and use 'ffffffffffffffffffffff' as preview text, the text selection over there goes quirky as well, you can only select 2 characters at the same time.

I have had this problem before with other fonts. if this font is important for you, you can report this as a bug to the designer. In this case: https://github.com/vernnobile/PacificoFont. Most designers with fonts featured on Google are willing to help. If they fix it, and you are using the font directly from Google, you don't have to do anything. If you downloaded the .ttf and are hosting it yourself, you should re-download it.

**Comment by asturur** [comment]
so fabric measures f and measures ff.

Then it says, when f is alone, is wide X, while if f is followed by f, is big as  ff minus the width of a singe f.

for fabric f and ff looks like this:
![image](https://user-images.githubusercontent.com/1194048/37823797-d89872d8-2e8a-11e8-8690-9329ac2dd2a4.png)

while in the specific case of a canvas in the demo:

![image](https://user-images.githubusercontent.com/1194048/37823848-10e907ba-2e8b-11e8-8f7c-2d4f98bf9fe2.png)

I have no idea why the measurement does not match the pixels

**Comment by blobinabottle** [comment]
It seems to be the case in all fabric apps. Even in shutterstock editor ;-) 
![image](https://user-images.githubusercontent.com/8090162/38562818-bb7635c8-3cdb-11e8-89fc-ced92c68f818.png)

Do you have an idea? The measures of `f ` vs `ff` are ok ?

**Comment by asturur** [comment]
i did not deep dive into it since is just Pacifico.
I m very stuck with work and clipPaths and would like to concentrate on those and bugfixes.

I have no means to check if what the browser says about f and ff is right. It seems right since one is smaller of the other.

I wonder what number we get measuring the whole string in a single shot.

**Comment by blobinabottle** [comment]
I'll check if I can find something that could help :-)

But clipPaths are definitely more interesting and needed than this 😃

**Comment by Mat-thieu** [comment]
The problem is still here and it can be really imprecise on many different types of fonts, take open sans here for example;

<img width="415" alt="Screenshot 2020-05-14 at 17 29 47" src="https://user-images.githubusercontent.com/7059091/81953935-8a27f280-9608-11ea-94f0-dcf90096fd5f.png">

It's provided by google fonts in this case.

I've been taking a look at the font measure functions and it's pretty hard to get into... maybe something to look out for is there possibly being floating point errors which are being added up for each character. While measuring the text it's constantly looking at the previous character and using it in its calculation for the next character. So if there's any error in the calculation it will get worse and worse the larger the text gets, and this is what's happening to the TextBox from what I can see

**Comment by asturur** [comment]
I'll try to describe how i built up text measurement in order to preserve kerning.

- text get measure at font size 400, to avoid small rounding of floats to have an impact
- text is splitted in grapheme clusters, and each grapheme cluster is measured with the previous one.

Example take the word `Vocal` with font pacifico, bold.

we create a key, `PACIFICTO_NORMAL_BOLD` ( because style, is not italic but normal )

then we measure V, because is the first letter, and we get a width. we save it in `PACIFICTO_NORMAL_BOLD.V`

Then we measure o, with previous character V.

We measure `Vo` and we save it in `PACIFICTO_NORMAL_BOLD.Vo`, then we subtract the size of V and we know the real size of `o` in this particular word/situation/sequence. We also save the size of `o` to reuse it later.

we move forward with c.
We measure `c` and `oc` to repeat the subtraction, and get the size of `c` alone and in this particular context.

Some font do not play well with this strategy, while a lot other do.

@Mat-thieu in the case of your open sans with the `d d  d ` what was exactly the string you were measuring?

**Comment by McFck** [comment]
Hello, is there any update to this issue? I still notice this exact problem with any font that has ligatures. Also I've seen shutterstock mentioned in this thread, but their canvas doesn't have this issue anymore, are they still using fabric ?

**Comment by asturur** [comment]
No they moved to a prebuilt platform with a custom engine that uses C code to measure text.
And it doesn't have this problem because ligatures have been disabled because caused issues.
So you shouldn't see ligatures working there as far as i remember

**Comment by asturur** [comment]
What is the exact font and string that is causing you issues?

**Comment by McFck** [comment]
I've tried both 'Pacifico' and 'Lab Grotesque'. In both scenarios string like  'fffffffffffffffffffff' will have a noticeable shift of cursor like this. And this problem only visible on this ligature for some reason.

![Image](https://github.com/user-attachments/assets/72412a9d-44cd-4f05-9ed3-366535e13366)

![Image](https://github.com/user-attachments/assets/7aee7a5e-cf35-43f4-bce1-699f0db793ef)

**Comment by asturur** [comment]
The issue is definitely related to fonts that change dramatically the size of drawings when certain combinations are hit, for example material-symbols is unusable as today.

Where the cursor sits in something like google docs?
Also what version of fabric are you using, before we debug an issue from 5.x or before.

**Comment by McFck** [comment]
fabric 6.6.1. Google docs seems to have it displayed correctly. Also editors like the one in the shutterstock that use webgl seems to have caret in the correct position

![Image](https://github.com/user-attachments/assets/ba614d81-1215-4358-a69c-d5e5ece512e4)

**Comment by asturur** [comment]
The shutterstock one works ok because ligatures have been disabled in pango when rendering text with plain cpu code and then uploading it as a texture. If you inspect closely you notice that the ffff of pacifico are not as near as the one in google docs.

**Comment by McFck** [comment]
Hmm, right. The skia one is this one https://holst.so, though it needs an account, but it's free.

**Comment by asturur** [comment]
ahhh is there an english link?

**Comment by McFck** [comment]
Yeah, just go to login https://app.holst.so/login it should be in English

**Comment by asturur** [comment]
<img width="776" alt="Image" src="https://github.com/user-attachments/assets/98586642-7558-4faa-b3f6-fd663b8fdd82" />

i see the basic behaviour is indeed better, and it start accumulating issues with ligatures as soon as you try styling
Thing that doesn't happen with a plain font.

this is not super straightforward to solve.
We could expand our cluster measuring from 2 to 3 characters, trying to understand how big a char is when is alone, when is followed, and so on, or maybe there is some specific issue in our logic that is triggered just by pacifico and similars.

Today what we do is that we measure first an `f`, the we measure the following one, `ff` and we say, ok the f is x and ff is y.
Then we memorize this and for each other ff we don't check anymore.
so we know that an f followed by another f is basically as big as y-x.

We could say, and that could be completely wrong, for couples of letters that are the same, the space is divided equally.

If we had access to internal font information from the api probably it would be easier

**Comment by McFck** [comment]
yeah, I've seen someone try to fix this using openType.js to read some font data. I've tried it, but for some reason it failed to read some of the data so I just skipped it
